boot/freeldr/freeldr/arch/uefi/uefidisk.c

```diff
@@ -0,0 +1,589 @@
+        return;
+    }
+
+    Buffer = (ULONG*)DiskReadBuffer;
+    Mbr = (PMASTER_BOOT_RECORD)DiskReadBuffer;
+
+    Signature = Mbr->Signature;
+    TRACE("Signature: %x\n", Signature);
+
+    /* Calculate the MBR checksum */
+    Checksum = 0;
+    for (i = 0; i < 512 / sizeof(ULONG); i++)
+    {
+        Checksum += Buffer[i];
+    }
+    Checksum = ~Checksum + 1;
+    TRACE("Checksum: %x\n", Checksum);
+
+    ValidPartitionTable = (Mbr->MasterBootRecordMagic == 0xAA55);
+
+    /* Fill out the ARC disk block */
+    sprintf(ArcName, "multi(0)disk(0)rdisk(%u)", DriveNumber - FIRST_BIOS_DISK);
+    AddReactOSArcDiskInfo(ArcName, Signature, Checksum, ValidPartitionTable);
+
+    sprintf(ArcName, "multi(0)disk(0)rdisk(%u)partition(0)", DriveNumber - FIRST_BIOS_DISK);
+    FsRegisterDevice(ArcName, &UefiDiskVtbl);
+
+    /* Add partitions */
+    i = FIRST_PARTITION;
+    DiskReportError(FALSE);
+    while (DiskGetPartitionEntry(DriveNumber, i, &PartitionTableEntry))
+    {
+        if (PartitionTableEntry.SystemIndicator != PARTITION_ENTRY_UNUSED)
+        {
+            sprintf(ArcName, "multi(0)disk(0)rdisk(%u)partition(%lu)", DriveNumber - FIRST_BIOS_DISK, i);
+            FsRegisterDevice(ArcName, &UefiDiskVtbl);
+        }
+        i++;
+    }
+    DiskReportError(TRUE);
+
+    InternalUefiDisk[DriveNumber].NumOfPartitions = i;
+    /* Convert checksum and signature to identifier string */
+    Identifier[0] = Hex[(Checksum >> 28) & 0x0F];
+    Identifier[1] = Hex[(Checksum >> 24) & 0x0F];
+    Identifier[2] = Hex[(Checksum >> 20) & 0x0F];
+    Identifier[3] = Hex[(Checksum >> 16) & 0x0F];
+    Identifier[4] = Hex[(Checksum >> 12) & 0x0F];
+    Identifier[5] = Hex[(Checksum >> 8) & 0x0F];
+    Identifier[6] = Hex[(Checksum >> 4) & 0x0F];
+    Identifier[7] = Hex[Checksum & 0x0F];
+    Identifier[8] = '-';
+    Identifier[9] = Hex[(Signature >> 28) & 0x0F];
+    Identifier[10] = Hex[(Signature >> 24) & 0x0F];
+    Identifier[11] = Hex[(Signature >> 20) & 0x0F];
+    Identifier[12] = Hex[(Signature >> 16) & 0x0F];
+    Identifier[13] = Hex[(Signature >> 12) & 0x0F];
+    Identifier[14] = Hex[(Signature >> 8) & 0x0F];
+    Identifier[15] = Hex[(Signature >> 4) & 0x0F];
+    Identifier[16] = Hex[Signature & 0x0F];
+    Identifier[17] = '-';
+    Identifier[18] = (ValidPartitionTable ? 'A' : 'X');
+    Identifier[19] = 0;
+    TRACE("Identifier: %s\n", Identifier);
+}
+
+static
+VOID
+UefiSetupBlockDevices()
+{
+
+    ULONG BlockDeviceIndex;
+    ULONG SystemHandleCount;
+    EFI_STATUS Status;
+    ULONG i;
+
+    UINTN handle_size = 0;
+    PcBiosDiskCount = 0;
+    UefiBootRootIdentifier = 0;
+
+    /* 1) Setup a list of boothandles by using the LocateHandle protocol */
+    Status = GlobalSystemTable->BootServices->LocateHandle(ByProtocol, &bioGuid, NULL, &handle_size, handles);
+    handles = MmAllocateMemoryWithType((handle_size + (EFI_PAGE_SIZE * 2) * 2), LoaderFirmwareTemporary);
```

Comment:
Why the "times 2" ?

Reply:
Oops left over from a copy and paste :), fixed in next --amend

Reply:
Waiting for the fix :)